embassy-embedded-hal/src/flash/partition.rs

```diff
@@ -0,0 +1,150 @@
+use embassy_sync::blocking_mutex::raw::RawMutex;
+use embassy_sync::mutex::Mutex;
+use embedded_storage::nor_flash::{ErrorType, NorFlashError, NorFlashErrorKind};
+#[cfg(feature = "nightly")]
+use embedded_storage_async::nor_flash::{NorFlash, ReadNorFlash};
+
+/// A logical partition of an underlying shared flash
+///
+/// A partition holds an offset and a size of the flash,
+/// and is restricted to operate with that range.
+/// There is no guarantee that muliple partitions on the same flash
+/// operate on mutually exclusive ranges - such a separation is up to
+/// the user to guarantee.
+pub struct Partition<'a, M: RawMutex, T> {
+    flash: &'a Mutex<M, T>,
+    offset: u32,
+    size: u32,
+}
+
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error<T> {
+    Partition,
+    OutOfBounds,
+    Flash(T),
+}
+
+impl<'a, M: RawMutex, T> Partition<'a, M, T> {
+    /// Create a new partition
+    pub const fn new(flash: &'a Mutex<M, T>, offset: u32, size: u32) -> Self {
+        Self { flash, offset, size }
+    }
+}
+
+impl<T: NorFlashError> NorFlashError for Error<T> {
+    fn kind(&self) -> NorFlashErrorKind {
+        match self {
+            Error::Partition => NorFlashErrorKind::Other,
+            Error::OutOfBounds => NorFlashErrorKind::OutOfBounds,
+            Error::Flash(f) => f.kind(),
+        }
+    }
+}
+
+impl<M: RawMutex, T: ErrorType> ErrorType for Partition<'_, M, T> {
+    type Error = Error<T::Error>;
+}
+
+#[cfg(feature = "nightly")]
+impl<M: RawMutex, T: ReadNorFlash> ReadNorFlash for Partition<'_, M, T> {
+    const READ_SIZE: usize = T::READ_SIZE;
+
+    async fn read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Self::Error> {
+        if self.offset % T::READ_SIZE as u32 != 0 || self.size % T::READ_SIZE as u32 != 0 {
```

Comment:
I think it'd be better to check in `new` that offset/size are multiple of ERASE_SIZE. We can asume ERASE_SIZE is a multiple of READ_SIZE/WRITE_SIZE, so this covers read/write as well. This way we don't have to do checks at runtime.

Reply:
oh but then it needs `where T: NorFlash` in `new()`, so we can't support readonly flashes. 

Perhaps that's okay? Readonly flash is rare. Also, we could add support if someone needs it by adding `ReadonlyPartition` or something.